File: 0141-linked-list-cycle/0141-linked-list-cycle.cpp
/**  
 * Definition for singly-linked list.  
 * struct ListNode {  
 *     int val;  
 *     ListNode *next;  
 *     ListNode(int x) : val(x), next(NULL) {}  
 * };  
 */  
class Solution {  
public:  
    bool hasCycle(ListNode *head) {  
        if (head == nullptr) {  
            return false; // If the list is empty, return false  
        }  
        
        ListNode* tx = head;          // Pointer moving one step  
        ListNode* ox = head;          // Pointer moving two steps  

        while (ox != nullptr && ox->next != nullptr) { // Check if ox and ox->next are valid  
            tx = tx->next;            // Move tx one step  
            ox = ox->next->next;      // Move ox two steps  

            if (tx == ox) {           // Check if they meet  
                return true;          // Cycle detected  
            }  
        }  
        return false;                 // No cycle  
    }  
};

File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        f = head
        s = head
        flag = 0

        while s and f and f.next:
            if s is  f and flag:
                return True 
            flag = 1

            if s :
                s = s.next
            if f.next:
                f = f.next.next
        return False